src/schema/firefox-schemas-import.js

```diff
@@ -576,30 +578,18 @@ inner.isBrowserSchema = (_path) => {
   return schemaRegexes.some((re) => re.test(_path));
 };
 
-/**
- * Strip a null byte if it's the last character in a string.
- * */
-export function stripTrailingNullByte(str) {
-  if (str.indexOf('\u0000') === str.length - 1) {
-    return str.slice(0, -1);
-  }
-  return str;
-}
-
 export async function fetchSchemas({ inputPath, outputPath }) {
-  const tarball = fs.createReadStream(inputPath);
+  const openZip = util.promisify(yauzl.open);
+  const zipfile = await openZip(inputPath);
+  const openReadStream = util.promisify(zipfile.openReadStream.bind(zipfile));
+
   return new Promise((resolve, reject) => {
-    tarball
-      .pipe(new tar.Parse())
-      .on('entry', (entry) => {
-        // The updated tar library doesn't always strip null bytes from the
-        // end of strings anymore so we get to do that here.
-        const entryPath = stripTrailingNullByte(entry.path);
-        if (inner.isBrowserSchema(entryPath)) {
-          const filePath = path.join(outputPath, path.basename(entryPath));
-          entry.pipe(fs.createWriteStream(filePath));
-        } else {
-          entry.resume();
+    zipfile
+      .on('entry', async (entry) => {
+        if (inner.isBrowserSchema(entry.fileName)) {
+          const filePath = path.join(outputPath, path.basename(entry.fileName));
+          const readStream = await openReadStream(entry);
+          readStream.pipe(fs.createWriteStream(filePath));
```

Comment:
While re-reading this part once more, I started to wonder if we should explicitly await for the write streams we are creating here, to be sure that all data in the write streams has been completely flushed and the file closed.

It looks that in the previous version based on the tar file we were already doing this without explicitly waiting the write streams to emit the close event (and that part shouldn't be affected by this change).

have we already tried this branch on a bunch of real mozilla zip files to double-check that the schema are imported successfully as expected?
It would be good to try it once more before merging this patch on master (if we have not already done it) and it may actually help us to double-check if explicitly waiting for the write streams to be closed is actually necessary.

If it turns out that we need to await the write stream close events (or we just want to do it to be absolutely sure), I think that we may achieve it easily enough by collecting all the related promises and then wait for them to be resolved or rejected using something like:

```
const writeStreamsPromises = [];
return new Promise((resolve, reject) => {
   ...
   .on('entry', async (entry) => {
      ...
      const destFileStream = fs.createWriteStream(filePath);
      writeStreamsPromises.push(
        new Promise(resolve => destFileStream.on("close", resolve)
      );
      const readStream = await openReadStream(entry);
      readStream.pipe(destFileStream);
      ...
   .on('end', () => 
       Promise.all(writeStreamsPromises).then(resolve, reject);
       ...
```

Reply:
Oh, good points. I'll look into that more deeply tomorrow or just steal your code-changes.

I tried out the patch on all the recent schema imports - just never finished up fixing the tests so it does work in real-life.

Reply:
Fixed in 
bebea50

Reply:
:heart: thanks!